### DISCUSSION TITLE
kiali behind proxy oidc configuration

### DISCUSSION THREAD

**Discussion by VanherpeDries** [kiali/kiali#3691] [Q&A]
Description
I'm running a kiali server on a k8s 1.19, behind a ambassador proxy. I'm trying out the OpenID connection with Kiali. This Kiali instance is running on : https://kiali.k8s.local . But when i try to connect to the openid server it points to : https://kiali.k8s.local:20001 which can't be fount

Versions used
Kiali: v1.29

config: 
  auth:
    strategy: openid
    openshift: {}
    openid:
      client_id: "kiali"
      api_proxy: "https://kiali.k8s.local"
      issuer_uri: "https://openid.local/poam/oauth2"
      scopes: ["openid", "profile", "email", "groups"]

**Comment by jmazzitelli** [comment]
Read this and see if related : https://github.com/kiali/kiali/issues/3111

As a result of that issue, we added this setting which sounds like you might need to set: https://github.com/kiali/kiali-operator/blob/v1.29.1/deploy/kiali/kiali_cr.yaml#L779-L785

See also: https://kiali.io/documentation/latest/faq/#x-forwarded-port

**Reply by VanherpeDries** [reply]
First of all thanks for helping! 
I've tried and use multiple of these configurations : 
- i've tried using the documentation you've mentioned above
- i've tried using the forwarded port
- i've tried a combination of both 

i've managed to get a bit further in the oidc flow, the request is sent but i'm getting : `{"error":"Token is not valid or is expired","detail":"Unauthorized"}`. I've checked this with the OIDC guys of our corporation, they said it could be with kiali wrongly interpreting the response, since the OIDC provider gets a 200 response code on the POST request to request a code and token. 


Also i should add more information : 
- oidc provider: forgerock
- reverse proxy: ambassador

**Comment by israel-hdez** [comment]
@VanherpeDries I see you originally posted that you are using `api_proxy: "https://kiali.k8s.local"`. Is this your current configuration, or you removed it?

I'm wondering if ambassador is being used as a reverse proxy for serving Kiali, or if it's placed in front of the Kubernetes API.

**Reply by VanherpeDries** [reply]
well it's an ingress controller in the kube stack
and `api_proxy: "https://kiali.k8s.local"` is still in the config

**Reply by israel-hdez** [reply]
I think, as first step, you will need to remove the `api_proxy`. That configuration is for cases when the proxy is for the Kubernetes API, rather than an ingress - for example, when you use [kube-oidc-proxy](https://github.com/jetstack/kube-oidc-proxy) (see https://kiali.io/documentation/latest/configuration/authentication/openid/#_set_up_with_rbac_support).

Try seeing if you can login with that configuration removed.

**Reply by VanherpeDries** [reply]
still getting the same error here is my config : 
````
  auth:
    strategy: openid
    openshift: {}
    openid:
      client_id: "kiali"
      client_secret: "****"
      issuer_uri: "****"
      scopes: ["openid", "profile", "email", "groups"]
  server:
    metrics_enabled: true
    metrics_port: 9090
    port: 20001
    web_port: 443
    web_fqdn: kiali.k8s.local
    web_schema: "https"
```

**Reply by israel-hdez** [reply]
Are you using a cloud provider, bare metal, something else?

**Reply by VanherpeDries** [reply]
its a k8s cluster on Red Hat Openstack

**Reply by israel-hdez** [reply]
I guess it qualifies as something similar to bare metal. I guess you installed the cluster using `kubeadm` right?

**Reply by VanherpeDries** [reply]
kubespray/kubeadm

**Reply by israel-hdez** [reply]
Please, check that  your Kubernetes cluster has the OIDC integration enabled.
This StackOverflow answer could give a hint about how to check if that config is enabled: https://stackoverflow.com/a/62908645

Indeed, the `{"error":"Token is not valid or is expired","detail":"Unauthorized"}` is the typical one I've seen when OIDC integration is not configured in the cluster, or when there is a config mismatch.

I suggest you to disable RBAC in Kiali to check that there is no other issue. See: https://kiali.io/documentation/latest/configuration/authentication/openid/#_set_up_with_no_rbac_support. If you can authenticate with RBAC turned off, IMO, the issue is because of configuration.

See the official k8s docs here to learn a little more: https://kubernetes.io/docs/reference/access-authn-authz/authentication/#openid-connect-tokens

If OIDC integration is enabled in your cluster, please check that the `client_id` that you have configured in Kiali is exactly the same as the client id that you configured in Kubernetes. I see that you have `client_id: "kiali"` for Kiali. Check that it matches what you have for kubespray/kubeadm.

**Reply by VanherpeDries** [reply]
We are already using the OIDC integration (dex) on the kubernetes and that seems to work fine, even when i turn off `rbac support ` it still gives the same error, i've also checked the kiali client_id & other OIDC connection parameters

**Reply by israel-hdez** [reply]
With RBAC turned off, you cannot get the `{"error":"Token is not valid or is expired","detail":"Unauthorized"}`. The message should be different and more detailed when RBAC is turned off. So, please, double check this with RBAC turned off.

Also, make sure you restart the Kiali pod after doing config changes, because config is not refreshed without a restart.

**Reply by matthiasdeblock** [reply]
Hi

I'm working together with Dries on this and I got a bit closer to the finish line but still running into some issues. 

First of all, when using the below config, I get the error that I am unable to view namespaces . With this I think I am some steps further then we where before with the unauthorized error (invalid or expired token):
```
    auth:
      strategy: openid
      openid:
        client_id: k8s
        issuer_uri: https://ISSUER_URL
        username_claim: "sub"
```
Error message when clicking the OpenID button:
```
"Cannot view any namespaces. Please read Kiali's RBAC documentation for more details."
```

Now, as we are using groups to login to k8s with dex authenticator I've added the following option to the config right below username_claim:
```
scopes: ["openid","email","profile","groups"]
```
But when I try to log in now I get the following error in the kiali logs:
```
2021-03-05T16:09:17Z INF Server endpoint will start at [:20001/kiali]
2021-03-05T16:09:17Z INF Server endpoint will serve static content from [/opt/kiali/console]
2021-03-05T16:09:17Z INF Starting Metrics Server on [:9090]
2021-03-05T16:10:44Z INF Not handling OpenId code flow authentication: No nonce code present. Login window timed out.
2021-03-05T16:10:46Z WRN Configured OpenID provider informs some of the configured scopes are unsupported. Users may not be able to login.
2021-03-05T16:10:47Z DBG Rest perf config QPS: 175.000000 Burst: 200
2021-03-05T16:10:47Z TRC [Kiali Cache] cacheIstioTypes map[AuthorizationPolicy:true DestinationRule:true Gateway:true PeerAuthentication:true RequestAuthentication:true ServiceEntry:true Sidecar:true VirtualService:true]
2021-03-05T16:10:47Z INF Kiali Cache is active for namespaces [.*]
2021-03-05T16:10:47Z DBG Rest perf config QPS: 175.000000 Burst: 200
2021-03-05T16:10:47Z INF [Prom Cache] Enabled
2021-03-05T16:10:47Z INF Not handling OpenId code flow authentication: No nonce code present. Login window timed out.
```
The most important line I think here is: 
```
Configured OpenID provider informs some of the configured scopes are unsupported. Users may not be able to login.
```
My browser is just bringing me back to the screen with the OpenID button. No error like before without the scopes.

The clusterrolebinding I used:
```
kubectl create clusterrolebinding kiali-admin --clusterrole=kiali --group=CN=gl_aa_kubernetes_admins,OU=KUBERNETES,OU=TOEPASSINGEN,OU=GROEPEN,OU=......
```
The groups is exactly the same group as I've used for the dex authentication with kubernetes (JWT).

The secret of the OIDC client is also mounted in the Pod.

Any thoughts? 

Thank you!
Regards
Matthias

**Reply by israel-hdez** [reply]
Can you check the browser's developer console?

I think Dex may be passing an error to Kiali, but Kiali is not showing the error in the browser. In the developer console of the browser you probably may see the redirection from dex to Kiali with the error in the query string.

Additionally, by seeing the format of your group name, I assume you are using LDAP. Although I'm no expert in Dex, I'm wondering if you have configured the `groupSearch` in Dex. I'm seeing this sample yaml: https://github.com/dexidp/dex/blob/f7d1405cfd91528a95ccfe9476c9fd9b6a991c2f/examples/ldap/config-ldap.yaml#L49-L60. I'm not sure if Dex will be able to report groups if the group search is not setup.

Dex's sample application may also help in debugging if Dex is rejecting the requests when using the `groups` scope: https://dexidp.io/docs/getting-started/#running-a-client

**Reply by matthiasdeblock** [reply]
Hi

For the record, we are using ForgeRock as issuer. Not Dex. 
For K8s we are using DEX-K8S-AUTHENTICATOR which generate token for our k8s (https://github.com/mintel/dex-k8s-authenticator).
When we issue a token, ForgeRock provides groups (same format as in the clusterrolebinding) and then we have clusterrolebindings with groups (same as in the clusterrolebinding) to provide access to k8s. 

Kiali isn't using the dex-k8s-authenticator. We do are using the same ForgeRock client as we use for dex-k8s-authenticator and the OIDC from kubernetes (api). 

The developer tools don't show any error.

Thank you!
Regards
Matthias

**Reply by israel-hdez** [reply]
Hi, again,

Perhaps, try this URL in your browser?:

`https://$your_issuer_url/.well-known/openid-configuration`

It should return a JSON document. On that JSON document, you should be able to find a `scopes_supported` attribute listing the scopes that clients can request. Verify that the list contains the `groups` scope that you are adding in Kiali config (or any other scope you are adding). If it's not there, it is quite likely that your IdP is simply rejecting the request because of the invalid scope.

**Reply by matthiasdeblock** [reply]
Hi

Indeed, the scope isn't in the list.

I've just tried to disable rbac and that seems to be working. I am able to login now but if I am correct now everyone can login.

So I am still wondering if it is possible to log in with the clusterrolebinding with matching groups as we do in kubernetes right now. This is an example token we receive from dex-k8s-authenticator we have to get before we can log in with kubectl to the cluster. Depending on the group you are in you get admin rights or not.
```
{
  "sub": "mdeblock",
  "tokenName": "id_token",
  "groups": [
    "CN=gl_aa_kubernetes_admins,OU=KUBERNETES,...,DC=local"
  ],
  "given_name": "Matthias",
  "aud": "k8s",
  "tokenType": "JWTToken",
}
```
_filtered out some fields in the above token_

Thank you!
Regards
Matthias

**Reply by israel-hdez** [reply]
Hi @matthiasdeblock 

I'd say that if _dex-k8s-authenticator_ is retrieving the groups correctly, it's a matter of replicating your _dex-k8s-authenticator_ config into Kiali. Without entering into deep details, the way things work in Kiali is that once the IdP has issued a token to Kiali, Kiali is simply passing the received token to Kubernetes to let it to directly do privileges checks -- Kiali is not really doing any privilege check.

So, try configuring into Kiali the very same scopes as in _dex-k8s-authenticator_. It will be very weird if the IdP is issuing tokens with different data if configurations are the same. There should be some difference.

**Reply by matthiasdeblock** [reply]
Hi

I've put the exact same config for kiali as I did for the dex-k8s-authenticator but no progress. When I click the Login button Kiali redirects me to the ForgeRock, after signing in I get redirected back to the Login button page. 

dex-k8s-authenticator:
```
https://issuer_url:443/poam/oauth2/authorize?access_type=offline&client_id=k8s&redirect_uri=https%3A%2F%2Fdex-k8s-authenticator.k8s.cluster.local%2Fcallback&response_type=code&scope=openid+profile+email+offline_access+groups&state=Vgn2lp5QnymFtLntKX5dM8k773PwcM87T4hQtiESC1q8wkUBgw5D3kH0r5qJ
```
Kiali:
```
https://issuer_url:443/poam/oauth2/authorize?client_id=k8s&response_type=code&redirect_uri=https%3A%2F%2Fkiali.k8s.cluster.local%2Fkiali&scope=openid+email+profile+offline_access+groups&nonce=77a21ddd015a4a11ce24fe9309ef9bb53b641169b5493ee5893508fd&state=6a66e3ec858dc8f3eb2fda98d6ded3548ecb2730742dc86bfe514df0-210316134214
```

The info response (network)
```
strategy | "openid"
-- | --
authorizationEndpoint | "https://kiali.k8s.cluster.local/kiali/api/auth/openid_redirect"
sessionInfo | Object {  }
```

I do notice that there is 301 (Moved Permanently) response followed by a 302 after the request to the issuer. So, it is going from /kiali to /kiali/. 

Logs from Kiali pod:
```
2021-03-16T13:56:32Z DBG Rest perf config QPS: 175.000000 Burst: 200
2021-03-16T13:56:32Z INF Not handling OpenId code flow authentication: No nonce code present. Login window timed out.
```

Regards
Matthias

**Reply by israel-hdez** [reply]
@matthiasdeblock I though Kiali was showing a `Cannot view any namespaces. Please read Kiali's RBAC documentation for more details.` error. That was what I read in previous comments.

So, if Kiali is showing the login page without any error, I think you are right that the redirection may be the issue.
The issue in on our radar. It's already reported in #3762.

As a workaround, I'd suggest you to try another browser.

**Reply by matthiasdeblock** [reply]
Hi

That was when I only used the scopes openid, email and profile. When I added groups I got redirected as well as mentioned earlier. And indeed, It seems like it's the same behavior as in ticket #3762.

I did try your workaround but still nothing. Tried with 3 different browsers on 2 different VMs.

FYI: I've upgraded from 1.29 to 1.31 this afternoon. 

Regards
Matthias
